File: 289.game-of-life.py
#
# @lc app=leetcode id=289 lang=python3
#
# [289] Game of Life
#

# @lc code=start
from functools import cache
from itertools import product


class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        """
        og new
         0  0:0
         1  1:0
         0  2:1
         1  3:1

        """
        n, m = len(board), len(board[0])

        def survives(r, c):
            cnt = 0
            lc, rc = c > 0, c < m - 1
            if r > 0:
                lo = board[r - 1]
                cnt += (lo[c] % 2) + (lc and lo[c - 1] % 2) + (rc and lo[c + 1] % 2)
            if r < n - 1:
                hi = board[r + 1]
                cnt += (hi[c] % 2) + (lc and hi[c - 1] % 2) + (rc and hi[c + 1] % 2)
            cnt += (lc and board[r][c - 1] % 2) + (rc and board[r][c + 1] % 2)
            return cnt in (2, 3) if board[r][c] == 1 else cnt == 3

        for r, c in product(range(n), range(m)):
            if survives(r, c):
                board[r][c] |= 2
        for r, c in product(range(n), range(m)):
            board[r][c] >>= 1


# @lc code=end

File: RustSolutions/src/289.game-of-life.rs
/*
 * @lc app=leetcode id=289 lang=rust
 *
 * [289] Game of Life
 */

// @lc code=start
impl Solution {
    pub fn game_of_life(board: &mut Vec<Vec<i32>>) {
        let n = board.len();
        let m = board[0].len();

        fn count_neighbours(board: &Vec<Vec<i32>>, n: usize, m: usize, r: usize, c: usize) -> i32 {
            let coords = vec![
                (r - 1, c - 1),
                (r - 1, c),
                (r - 1, c + 1),
                (r, c - 1),
                (r, c + 1),
                (r + 1, c - 1),
                (r + 1, c),
                (r + 1, c + 1),
            ];

            coords.iter()
                .filter(move |&&(nr, nc)| nr < n && nc < m && board[nr][nc] % 2 == 1)
                .count() as i32
        }


        for r in 0..n {
            for c in 0..m {
                let count = count_neighbours(&board,n,m,r, c);
                if (board[r][c] == 1 && (count == 2 || count == 3)) || (board[r][c] == 0 && count == 3) {
                    board[r][c] |= 2;
                }
            }
        }

        for r in 0..n {
            for c in 0..m {
                board[r][c] >>= 1;
            }
        }
    }
}


// @lc code=end

